examples/cis2-wccd-upgrade/src/lib.rs

```diff
@@ -0,0 +1,2030 @@
+        amount: ContractTokenAmount,
+        owner: &Address,
+        state_builder: &mut StateBuilder<S>,
+    ) -> ContractResult<()> {
+        ensure_eq!(token_id, &TOKEN_ID_WCCD, ContractError::InvalidTokenId);
+        let mut owner_state = self.token.entry(*owner).or_insert_with(|| AddressState {
+            balance:   0u64.into(),
+            operators: state_builder.new_set(),
+        });
+
+        owner_state.balance += amount;
+        Ok(())
+    }
+
+    fn burn(
+        &mut self,
+        token_id: &ContractTokenId,
+        amount: ContractTokenAmount,
+        owner: &Address,
+    ) -> ContractResult<()> {
+        ensure_eq!(token_id, &TOKEN_ID_WCCD, ContractError::InvalidTokenId);
+        if amount == 0u64.into() {
+            return Ok(());
+        }
+
+        let mut from_state = self.token.get_mut(owner).ok_or(ContractError::InsufficientFunds)?;
+        ensure!(from_state.balance >= amount, ContractError::InsufficientFunds);
+        from_state.balance -= amount;
+
+        Ok(())
+    }
+}
+
+// Contract functions
+
+/// Initialize the state contract instance with no initial tokens.
+/// Logs a `Mint` event for the single token id with no amounts.
+/// TODO: move event logs to proxy
+#[init(contract = "CIS2-wCCD-State", enable_logger)]
+fn contract_state_init<S: HasStateApi>(
+    ctx: &impl HasInitContext,
+    state_builder: &mut StateBuilder<S>,
+    logger: &mut impl HasLogger,
+) -> InitResult<State<S>> {
+    // Construct the initial contract state.
+    let state = State::new(state_builder);
+    // Get the instantiater of this contract instance.
+    let invoker = Address::Account(ctx.init_origin());
+    // Log event for the newly minted token.
+    logger.log(&Cis2Event::Mint(MintEvent {
+        token_id: TOKEN_ID_WCCD,
+        amount:   ContractTokenAmount::from(0u64),
+        owner:    invoker,
+    }))?;
+
+    // Log event for where to find metadata for the token
+    logger.log(&Cis2Event::TokenMetadata::<_, ContractTokenAmount>(TokenMetadataEvent {
+        token_id:     TOKEN_ID_WCCD,
+        metadata_url: MetadataUrl {
+            url:  String::from(TOKEN_METADATA_URL),
+            hash: None,
+        },
+    }))?;
+
+    Ok(state)
+}
+
+/// Initialzes the state of the w_ccd contract with the proxy and the
+/// implementation addresses.
+#[receive(
+    contract = "CIS2-wCCD-State",
+    name = "initialize",
+    parameter = "InitializeStateParams",
+    mutable
+)]
+fn contract_state_initialize<S: HasStateApi>(
+    ctx: &impl HasReceiveContext,
+    host: &mut impl HasHost<State<S>, StateApiType = S>,
+) -> ContractResult<()> {
+    ensure!(!host.state().initialized, concordium_cis2::Cis2Error::Custom(AlreadyInitialized));
+    // Set proxy and implementation addresses.
+    let params: InitializeStateParams = ctx.parameter_cursor().get()?;
+    host.state_mut().proxy_address = params.proxy_address;
+    host.state_mut().implementation_address = params.implementation_address;
```

Comment:
Does it ever make sense to initilize this with `None`?

If not, I recommend removing the boolean flag alltogether and using `proxy_address.is_some()` as "is initialized" status.